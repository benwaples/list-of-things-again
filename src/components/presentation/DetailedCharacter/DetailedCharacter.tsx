import React from 'react'
import { useCharacterById } from '../../../hooks'

export const DetailedCharacter = ({id}: {id: string}) => {
  const {loading, error, character} = useCharacterById(id)
  console.log(id)
  if(loading) return <h1>Loading...</h1>
  if(error) return <h1>Failed to get more info on character</h1>

  return (
    <div>
      <h1>{character.name}</h1>
      <figure>
        <img src={character.imageUrl} alt={character.name} />
        <figcaption>{`Character id: ${character.id}`}</figcaption>
      </figure>
    </div>
  )
}
