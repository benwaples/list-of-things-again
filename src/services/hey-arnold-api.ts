type CharacterShape = {
  _id: string, 
  name: string, 
  image: string
}

export type Character = {
  id: string, 
  name: string, 
  imageUrl: string
}

export const getCharacters = (): Promise<Character[]> => {
  return fetch('https://hey-arnold-api.herokuapp.com/api/v1/characters')
    .then(res => res.json())
    .then(json => json.map((character: CharacterShape) => ({
      id: character._id,
      name: character.name,
      imageUrl: character.image
    })))
}

export const getCharacterById = (id: string): Promise<Character> => {
  return fetch(`https://hey-arnold-api.herokuapp.com/api/v1/characters/${id}`)
    .then(res => res.json())
    .then((character: CharacterShape) => ({
      id: character._id,
      name: character.name,
      imageUrl: character.image
    }))
}