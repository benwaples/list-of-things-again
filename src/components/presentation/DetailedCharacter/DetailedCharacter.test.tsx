import React from 'react'
import {
  screen,
  render,
  waitFor
} from '@testing-library/react'
import { getCharacterById } from '../../../services/hey-arnold-api'
import { act } from 'react-dom/test-utils'
import { MemoryRouter } from 'react-router-dom'
import { DetailedCharacter } from './DetailedCharacter'

jest.mock('../../../services/hey-arnold-api')

describe('DetailedCharacter Component behavior test', () => {
  it('should render a list of characters', () => {
    act(() => {
      getCharacterById.mockResolvedValue({
          id: 'my', 
          name: 'name', 
          imageUrl: 'is'
        })
    })

    render(<MemoryRouter>
      <DetailedCharacter id={'kasjdf;kljasd'}/>
    </MemoryRouter>)

    screen.getByText('Loading...')

    return waitFor(() => {
      expect(screen.getByText('name')).not.toBeEmptyDOMElement()
    })
  })

  it('should look the same as now with a snapshot test', () => {
    act(() => {
      getCharacterById.mockResolvedValue({
          id: 'my', 
          name: 'name', 
          imageUrl: 'is'
        })
    })

    const { asFragment } = render(<MemoryRouter>
      <DetailedCharacter id={';lkasdjf'} />
    </MemoryRouter>)

    return waitFor(() => {
      // eslint-disable-next-line testing-library/no-wait-for-snapshot
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
