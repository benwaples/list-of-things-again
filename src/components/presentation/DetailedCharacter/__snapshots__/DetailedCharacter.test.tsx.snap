// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailedCharacter Component behavior test should look the same as now with a snapshot test 1`] = `
<DocumentFragment>
  <h1>
    Loading...
  </h1>
</DocumentFragment>
`;

exports[`DetailedCharacter Component behavior test should look the same as now with a snapshot test 2`] = `
<DocumentFragment>
  <div>
    <h1>
      name
    </h1>
    <figure>
      <img
        alt="name"
        src="is"
      />
      <figcaption>
        Character id: my
      </figcaption>
    </figure>
  </div>
</DocumentFragment>
`;
